File: 101/Solution.java
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root.left);
        q.offer(root.right);
        while (!q.isEmpty()) {
            TreeNode left = q.poll();
            TreeNode right = q.poll();
            if (left == null && right != null || left != null && right == null) {
                return false;
            } else if (left == null && right == null) {
                continue;
            } else if (left.val != right.val) {
                return false;
            } else {
                q.offer(left.left);
                q.offer(right.right);
                q.offer(left.right);
                q.offer(right.left);
            }
        }
        return true;
    }
}

File: 101/solution.go
package _101

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type QueueItem struct {
	Val *TreeNode
	Nxt *QueueItem
}

type Queue struct {
	Head *QueueItem
	Tail *QueueItem
}

func (q *Queue) append(node *TreeNode) {
	item := &QueueItem{node, nil}
	if q.Head == nil {
		q.Head = item
		q.Tail = item
	} else {
		q.Tail.Nxt = item
		q.Tail = q.Tail.Nxt
	}
}

func (q *Queue) popLeft() *TreeNode {
	if q.IsEmpty() {
		return nil
	}
	res := q.Head.Val
	q.Head = q.Head.Nxt
	if q.Head == nil {
		q.Tail = nil
	}
	return res
}

func (q *Queue) IsEmpty() bool {
	return q.Head == nil
}

func isSymmetric(root *TreeNode) bool {
	if root == nil {
		return true
	}
	queue := Queue{}
	queue.append(root.Left)
	queue.append(root.Right)
	for !queue.IsEmpty() {
		left := queue.popLeft()
		right := queue.popLeft()
		if left != nil && right == nil || left == nil && right != nil {
			return false
		} else if left == nil && right == nil {
			continue
		} else if left.Val != right.Val {
			return false
		} else {
			queue.append(left.Left)
			queue.append(right.Right)
			queue.append(left.Right)
			queue.append(right.Left)
		}
	}
	return true
}
